src/System.Private.CoreLib/shared/System/Diagnostics/Tracing/PollingCounter.cs

```diff
@@ -0,0 +1,200 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.Collections;
+using System.Collections.Generic;
+using System.Threading;
+#if ES_BUILD_PCL
+    using System.Threading.Tasks;
+#endif
+
+#if ES_BUILD_STANDALONE
+namespace Microsoft.Diagnostics.Tracing
+#else
+namespace System.Diagnostics.Tracing
+#endif
+{
+    /// <summary>
+    /// PollingCounter is a variant of EventCounter - it collects and calculates similar statistics 
+    /// as EventCounter. PollingCounter differs from EventCounter in that it takes in a callback
+    /// function to collect metrics on its own rather than the user having to call WriteMetric() 
+    /// every time.
+    /// </summary>
+    internal partial class PollingCounter : BaseCounter
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="PollingCounter"/> class.
+        /// PollingCounter live as long as the EventSource that they are attached to unless they are
+        /// explicitly Disposed.
+        /// </summary>
+        /// <param name="name">The name.</param>
+        /// <param name="eventSource">The event source.</param>
+        public PollingCounter(string name, EventSource eventSource, Func<float> getMetricFunction) : base(name, eventSource)
+        {
+            _min = float.PositiveInfinity;
+            _max = float.NegativeInfinity;
+            _getMetricFunction = getMetricFunction;
+
+            InitializeBuffer();
+        }
+
+        /// <summary>
+        /// Calls the callback function to get the value to enqueue. 
+        /// </summary>
+        public void UpdateMetric()
+        {
+            try
+            {
+                Enqueue(_getMetricFunction());    
```

Comment:
This code is taking on extra complexity to handle the case where it will poll more than once per emission interval, even though that should never occur. If you call _getMetricFunction() once inside GetCounterPayload() then you will not need any queues, buffers, or saved statistics.